### ISSUE TITLE
iOS - System.ExecutionEngineException: Attempting to JIT compile method

### ISSUE THREAD

**Issue by whitewolf33** [twilio/TwilioChatXamarinBindings#3]
Hi @aleksandrsivanovs 

After successfully getting the chatToken for the Identity and the TokenProvider URL (using Twilio Functions as recommended), an exception is thrown while trying to create the chat client with the token. 

Exception details:

> {System.ExecutionEngineException: Attempting to JIT compile method 'Twilio.Chat.iOS.TwilioChatClient:ChatClientWithToken (string,Twilio.Chat.iOS.TwilioChatClientProperties,Twilio.Chat.iOS.TwilioChatClientDelegate,Twilio.Chat.iOS.TwilioClientCompletion)' while running in aot-only mode. See https://developer.xamarin.com/guides/ios/advanced_topics/limitations/ for more information.
> 
>   at ChatDemo.iOS.TwilioChatHelper.CreateClient (System.String chatToken) [0x00011] in /github/TwilioChatXamarinBindings/Sample/ChatDemo.iOS/TwilioChatHelper.cs:65 }
> 	base: {System.SystemException}

Line of code where the exception is thrown:

https://github.com/twilio/TwilioChatXamarinBindings/blob/master/Sample/ChatDemo.iOS/TwilioChatHelper.cs#L63

**Note**: Please add the latest nuget package Twilio.Chat.Xamarin (0.1.1) to the ChatDemo.iOS project and remove the reference to the native binding project.

**Comment by whitewolf33** [comment]
I tried creating the instance of the TwilioChatClient and update the token, but had the same error thrown.

```
 public void CreateClient(string chatToken)
        {
            try
            {
                this.twilioChatClient = null;
                TwilioChatClient.LogLevel = LogLevel.Info;
                TwilioChatClient newChatClient = new TwilioChatClient();
                newChatClient.UpdateToken(chatToken, (result) =>
                {
                    if (result.IsSuccessful)
                    {
                        Logger.Info("TwilioChatHelper", "Client created");
                        this.RegisterForNotificationsFromTwilioChatService(newChatClient);
                        this.twilioChatClient = newChatClient;
                        newChatClient.ChannelsList
                                            .PublicChannelDescriptorsWithCompletion(HandleChannelDescriptorPaginatorCompletion);
                    }
                    else
                    {
                        Logger.Error("TwilioChatHelper", $"Can't create client: {result.Error.Description}");
                    }


                });
            }
            catch (Exception ex)
            {
                Logger.Info("Exception", ex);
            }
        }
```

**Comment by aleksandrsivanovs** [comment]
Hi!

I've tried to compile the sample with packages installed from nuget.org and ran it on actual device and in simulator - all works as it supposed to.

can you try to update your visual studio? (first time I'm seeing this error, however, some of the forums advices to upgrade the VS itself)

**Comment by whitewolf33** [comment]
Hi @aleksandrsivanovs 

I am already on the latest version. For the benifit of the doubt, I downloaded the sample again and without changing anything but removing the reference to the Twilio.Chat.iOS project and adding the reference to the nuget package 0.1.1 in the ChatDemo.iOS project, I am getting a 

> Method 'ObjCRuntime.BlockLiteral.SetupBlockUnsafe' not found.

 error. 

This error is not exactly the same as I had mentioned but it is in the same exact line.
https://github.com/twilio/TwilioChatXamarinBindings/blob/master/Sample/ChatDemo.iOS/TwilioChatHelper.cs#L63

I have attached the project as is and the error that I am getting below. 
Btw, I am on VS Mac 7.3.3 (build 12) - up to date - when I checked for updates. 

Please try the attached project and let me know if you still do not have any issues. Thanks.

[TwilioChatXamarinBindings-master.zip](https://github.com/twilio/TwilioChatXamarinBindings/files/1694734/TwilioChatXamarinBindings-master.zip)

Error Screen shot:

![screen shot 2018-02-05 at 9 39 22 pm](https://user-images.githubusercontent.com/4328043/35800599-89cb17b4-0abd-11e8-80f7-6d1f833e495e.png)


Stack Trace: 
>  at ChatDemo.iOS.TwilioChatHelper.CreateClient (System.String chatToken) [0x00010] in /Users/vinod/Desktop/TwilioChatXamarinBindings-master/Sample/ChatDemo.iOS/TwilioChatHelper.cs:63 
>   at ChatDemo.ChatDemoLoginPage+<OnLoginButtonClicked>d__1.MoveNext () [0x00267] in /Users/vinod/Desktop/TwilioChatXamarinBindings-master/Sample/ChatDemo/ChatDemoLoginPage.xaml.cs:57 
> --- End of stack trace from previous location where exception was thrown ---
>   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.6.1.4/src/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:152 
>   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.6.1.4/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1018 
>   at UIKit.UIKitSynchronizationContext+<Post>c__AnonStorey0.<>m__0 () [0x00000] in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/UIKit/UIKitSynchronizationContext.cs:24 
>   at Foundation.NSAsyncActionDispatcher.Apply () [0x00000] in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/Foundation/NSAction.cs:163 
> --- End of stack trace from previous location where exception was thrown ---
>   at (wrapper managed-to-native) UIKit.UIApplication:UIApplicationMain (int,string[],intptr,intptr)
>   at UIKit.UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate) [0x00005] in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/UIKit/UIApplication.cs:79 
>   at UIKit.UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName) [0x00038] in /Users/builder/data/lanes/5665/db807ec9/source/xamarin-macios/src/UIKit/UIApplication.cs:63 
>   at ChatDemo.iOS.Application.Main (System.String[] args) [0x00001] in /Users/vinod/Desktop/TwilioChatXamarinBindings-master/Sample/ChatDemo.iOS/Main.cs:17

**Comment by whitewolf33** [comment]
I tried to create a sample from scratch and created a Blank Forms App with PCL and tried to add the Twilio.Chat.Xamarin nuget to the PCL project, to have the interface ITwilioChatHelper in the PCL project. But looks like the package is not compatible with PCL Profile 111

Is this intentional that the nuget is not supporting the PCL or does the package support profile other than 111? Can you please let me know?

> Attempting to gather dependency information for package 'Twilio.Chat.Xamarin.0.1.1' with respect to project 'TwilioChatTest', targeting '.NETPortable,Version=v4.5,Profile=Profile111'
>   GET https://api.nuget.org/v3/registration3-gz-semver2/twilio.chat.xamarin/index.json
>   OK https://api.nuget.org/v3/registration3-gz-semver2/twilio.chat.xamarin/index.json 294ms
> Total number of results gathered : 2
> Gathering dependency information took 297.25 ms
> Summary of time taken to gather dependencies per source :
> https://api.nuget.org/v3/index.json	-	296.75 ms
> Attempting to resolve dependencies for package 'Twilio.Chat.Xamarin.0.1.1' with DependencyBehavior 'Lowest'
> Resolving dependency information took 0 ms
> Resolving actions to install package 'Twilio.Chat.Xamarin.0.1.1'
> Resolved actions to install package 'Twilio.Chat.Xamarin.0.1.1'
> Found package 'Twilio.Chat.Xamarin 0.1.1' in '/Users/vinod/Desktop/TwilioChatTest/packages'.
> For adding package 'Twilio.Chat.Xamarin.0.1.1' to project 'TwilioChatTest' that targets 'portable45-net45+win8+wpa81'.
> Install failed. Rolling back...
> Package 'Twilio.Chat.Xamarin.0.1.1' does not exist in project 'TwilioChatTest'
> Executing nuget actions took 15.12 ms
> Could not install package 'Twilio.Chat.Xamarin 0.1.1'. You are trying to install this package into a project that targets '.NETPortable,Version=v4.5,Profile=Profile111', but the package does not contain any assembly references or content files that are compatible with that framework. For more information, contact the package author.

**Comment by aleksandrsivanovs** [comment]
iirc, "Method 'ObjCRuntime.BlockLiteral.SetupBlockUnsafe' not found" is because you don't have latest xamarin.ios installed. about PCL - not intentional for sure, I have to investigate more.

**Comment by aleksandrsivanovs** [comment]
and, yes, compiled and run (and login and got to working state) the attached zipped solution without problems. smth is strange in the setup itself. frankly speaking, I don't know what can it be :(

**Comment by whitewolf33** [comment]
Hi @aleksandrsivanovs 

Let me try it on my other Mac and will report back.  Given that there was an issue with the nuget package for iOS, has anyone else tried installing it? Can you please try it on another machine apart from your Dev machine? I am also interested in the nuget package to be compatible with the PCL project (Should I raise a seperate issue to avoid distractions? Please let me know).  

Regarding the Xamarin version, I am sure I am running the latest stable version according to 

https://docs.microsoft.com/en-gb/visualstudio/releasenotes/vs2017-mac-relnotes

Xamarin iOS is 11.6.1.4 (11.8 is still in preview)
https://developer.xamarin.com/releases/ios/xamarin.ios_11/xamarin.ios_11.6/

System Info from my Xamarin Studio for Mac.

=== Visual Studio Community 2017 for Mac ===

Version 7.3.3 (build 12)
Installation UUID: eb28094d-a68e-44e2-971e-fe6aeb511439
Runtime:
	Mono 5.4.1.7 (2017-06/e66d9abbb27) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)

	Package version: 504010007

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.0.0
	1.1.2
	1.0.5
SDK: /usr/local/share/dotnet/sdk/2.0.0/Sdks
SDK Versions:
	2.0.0
	1.0.4
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.4.1/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.0
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Apple Developer Tools ===

Xcode 9.2 (13772)
Build 9C40b

=== Xamarin.iOS ===

Version: 11.6.1.4 (Visual Studio Community)
Hash: db807ec9
Branch: xcode9.2
Build date: 2018-01-10 16:45:48-0500

=== Xamarin.Android ===

Version: 8.1.5.0 (Visual Studio Community)
Android SDK: /Users/vinod/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		5.1 (API level 22)
		6.0 (API level 23)
		7.1 (API level 25)
		8.0 (API level 26)

SDK Tools Version: 25.2.5
SDK Platform Tools Version: 25.0.4
SDK Build Tools Version: 25.0.1

Java SDK: /usr
java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin.Mac ===

Version: 4.0.0.216 (Visual Studio Community)

=== Xamarin Inspector ===

Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 703030012
Git revision: b07492f1e48be596bad92dc4b7a3bc2d128ed0f9
Build date: 2018-01-30 13:15:55-05
Xamarin addins: 7c8f967d67207118dd99a1d0cc9c228045b30c5f
Build lane: monodevelop-lion-d15-5

=== Operating System ===

Mac OS X 10.13.2
Darwin 17.3.0 Darwin Kernel Version 17.3.0
    Thu Nov  9 18:09:22 PST 2017
    root:xnu-4570.31.3~1/RELEASE_X86_64 x86_64

=== Enabled user installed addins ===

Redth's Addins 1.0.9
Internet of Things (IoT) development (Preview) 7.1

**Comment by aleksandrsivanovs** [comment]
for PCL compatibility please raise another issue to track it separately.

as for my Xamarin.iOS:  11.8.0.8
seems like here is the main problem you are mentioning.
not sure how to downgrade, though.
will try and then will report back (this will take some time, though)

**Comment by whitewolf33** [comment]
I can confirm that the Sample project has got the Missing Method Exception.

I downloaded the sample project on another Mac (similar versioning - latest stable), and had the same issue, as mentioned above.

Method 'ObjCRuntime.BlockLiteral.SetupBlockUnsafe' not found.

**Comment by aleksandrsivanovs** [comment]
so, downgraded to:
```
=== Visual Studio Community 2017 for Mac ===

Version 7.3.3 (build 12)
Installation UUID: 0fd30394-bfb9-4eb8-8b3e-c03d61c5dca0
Runtime:
	Mono 5.4.1.7 (2017-06/e66d9abbb27) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)

	Package version: 504010007

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Version: 2.0.0
SDK: /usr/local/share/dotnet/sdk/2.0.0/Sdks
SDK Version: 2.0.0
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.4.1/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.0
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Apple Developer Tools ===

Xcode 9.2 (13772)
Build 9C40b

=== Xamarin.Mac ===

Version: 4.0.0.216 (Visual Studio Community)

=== Xamarin.Android ===

Version: 8.1.5.0 (Visual Studio Community)
Android SDK: /Users/aivanovs/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		4.0.3  (API level 15)
		4.4    (API level 19)
		4.4.87 (API level 20)
		5.1    (API level 22)
		6.0    (API level 23)
		7.0    (API level 24)
		7.1    (API level 25)

SDK Tools Version: 25.2.5
SDK Platform Tools Version: 25.0.5
SDK Build Tools Version: 25.0.3

Java SDK: /usr
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin Inspector ===

Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

=== Xamarin.iOS ===

Version: 11.6.1.4 (Visual Studio Community)
Hash: db807ec9
Branch: xcode9.2
Build date: 2018-01-10 16:45:48-0500

=== Build Information ===

Release ID: 703030012
Git revision: b07492f1e48be596bad92dc4b7a3bc2d128ed0f9
Build date: 2018-01-30 13:15:55-05
Xamarin addins: 7c8f967d67207118dd99a1d0cc9c228045b30c5f
Build lane: monodevelop-lion-d15-5

=== Operating System ===

Mac OS X 10.12.6
Darwin 16.7.0 Darwin Kernel Version 16.7.0
    Thu Jan 11 22:59:40 PST 2018
    root:xnu-3789.73.8~1/RELEASE_X86_64 x86_64

=== Enabled user installed addins ===

Internet of Things (IoT) development (Preview) 7.1
````

built new version of the package (0.1.2), deployed to local nuget store, installed from it and ran the app on standalone version - all is good.

the question is - I would like you to try it out first on your setup. 
however, I can't find good way to share new nupkg with you, how can we arrange this binary transfer? :)

**Comment by whitewolf33** [comment]
Tried the nuget package 0.1.2 with the chatdemo project for iOS and it is a winner. Can see the message logs for a given identity. Nice work!

Eagerly waiting for the PCL issue to be resolved to try it out. Will close this issue now.

![chatdemo_logs](https://user-images.githubusercontent.com/4328043/35858712-97ef40e8-0b91-11e8-9f9a-760982992226.jpg)

**Comment by aleksandrsivanovs** [comment]
awesome, I'll publish 0.1.2 to the nuget.org then.

thanks for report and cooperation :)

**Comment by aleksandrsivanovs** [comment]
published as 0.1.2: https://www.nuget.org/packages/Twilio.Chat.Xamarin/0.1.2
